torch/_inductor/optimize_indexing.py

```diff
@@ -0,0 +1,473 @@
+import dataclasses
+import functools
+import itertools
+import math
+import operator
+from typing import Dict, Iterable, Union
+
+import sympy
+
+import torch
+from .ir import IndexingDiv, InterpreterShim, LoopBody, ModularIndexing
+from .utils import sympy_subs
+from .virtualized import V
+
+
+@dataclasses.dataclass(frozen=True)
+class ValueRanges(object):
+    lower: Union[sympy.Symbol, sympy.Number, int, float, bool]
+    upper: Union[sympy.Symbol, sympy.Number, int, float, bool]
+
+    @classmethod
+    def wrap(cls, arg):
+        if isinstance(arg, ValueRanges):
+            return arg
+        assert isinstance(arg, (int, float, bool))
+        return ValueRanges(arg, arg)
+
+    @classmethod
+    def unary_map(cls, x, fn):
+        "map lower and upper bound with fn"
+        x = cls.wrap(x)
+        return ValueRanges(fn(x.lower), fn(x.upper))
+
+    @classmethod
+    def checked_unary_map(cls, x, fn):
+        "check the max and min of computed upper and lower bound for the output"
+        out = cls.unary_map(x, fn)
+        return ValueRanges(min(out.lower, out.upper), max(out.lower, out.upper))
+
+    @classmethod
+    def binary_map(cls, x, y, fn):
+        "map upper and lower bounds accessing corresponding values of inputs"
+        x, y = cls.wrap(x), cls.wrap(y)
+
+        return ValueRanges(
+            fn(x.lower, y.lower),
+            fn(x.upper, y.upper),
+        )
+
+    @classmethod
+    def binary_map_products(cls, a, b, fn):
+        "compute the product of all lower and upper bounds and take min and max"
+        a, b = cls.wrap(a), cls.wrap(b)
+        products = [
+            fn(x, y)
+            for x, y in itertools.product([a.lower, a.upper], [b.lower, b.upper])
+        ]
+        return ValueRanges(min(products), max(products))
+
+
+class ValueRangeAnalysis(object):
```

Comment:
I'd suggest adding a default handler to this class that returns `ValueRanges(-math.inf, math.inf)` and issues a warning in case we missed an op or a new op is added in the future.